src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs

```diff
@@ -112,95 +109,97 @@ public virtual Encoding SelectCharacterEncoding(OutputFormatterContext context)
+                // it want.
+                if (SupportedMediaTypes.Count > 0)
+                {
+                    context.ContentType = SupportedMediaTypes[0];
+                    return true;
+                }
+                else
+                {
+                    return false;
+                }
             }
             else
             {
                 // Confirm this formatter supports a more specific media type than requested e.g. OK if "text/*"
                 // requested and formatter supports "text/plain". contentType is typically what we got in an Accept
                 // header.
-                mediaType = SupportedMediaTypes.FirstOrDefault(
-                    supportedMediaType => supportedMediaType.IsSubsetOf(contentType));
-            }
-
-            if (mediaType != null)
-            {
-                context.SelectedContentType = mediaType;
-                return true;
+                for (var i = 0; i < SupportedMediaTypes.Count; i++)
+                {
+                    var mediaType = SupportedMediaTypes[i];
+                    if (mediaType.IsSubsetOf(context.ContentType))
+                    {
+                        context.ContentType = mediaType;
+                        return true;
+                    }
+                }
             }
 
             return false;
         }
 
         /// <inheritdoc />
-        public Task WriteAsync(OutputFormatterContext context)
-        {
-            if (context == null)
-            {
-                throw new ArgumentNullException(nameof(context));
-            }
-
-            WriteResponseHeaders(context);
-            return WriteResponseBodyAsync(context);
-        }
-
-        /// <summary>
-        /// Sets the headers on <see cref="Microsoft.AspNet.Http.HttpResponse"/> object.
-        /// </summary>
-        /// <param name="context">The formatter context associated with the call.</param>
-        public virtual void WriteResponseHeaders(OutputFormatterContext context)
+        public Task WriteAsync(OutputFormatterWriteContext context)
         {
             if (context == null)
             {
                 throw new ArgumentNullException(nameof(context));
             }
 
-            var selectedMediaType = context.SelectedContentType;
-
-            // If content type is not set then set it based on supported media types.
-            selectedMediaType = selectedMediaType ?? SupportedMediaTypes.FirstOrDefault();
+            var selectedMediaType = context.ContentType;
             if (selectedMediaType == null)
             {
-                throw new InvalidOperationException(Resources.FormatOutputFormatterNoMediaType(GetType().FullName));
+                // If content type is not set then set it based on supported media types.
+                if (SupportedEncodings.Count > 0)
+                {
+                    selectedMediaType = SupportedMediaTypes[0];
```

Comment:
This would happen if the user overrides `CanWriteResult` and returns true without setting a content-type. There is a phase in conneg where `context.ContentType` will be `null`.